### ISSUE TITLE
Issue with server rendering using Next.js

### ISSUE THREAD

**Issue by fabioespinosa** [react-querybuilder/react-querybuilder#128]
I'm having trouble server rendering this component using next.js.

I'm getting "self is not defined"

It's because `self` is only defined in the browser.

I wonder if you guys would consider a PR regarding a fix to this so we can enable react-querybuilder to be server-rendered? 👍 

Thanks for the great work.

**Comment by jakeboone02** [comment]
Absolutely we would consider that!

**Comment by jakeboone02** [comment]
Hi @fabioespinosa, I was wondering if you had had a chance to work on the server rendering problem.

**Comment by aftabnaveed** [comment]
I am using Express.js, the SSR is broken due to the `window.crypto` object being used. Can that not be encapsulated and avoided for SSR? Or at least replaced with a nodejs equivalent?

**Comment by jakeboone02** [comment]
@aftabnaveed that's a good idea -- I'll look into it.  I'd welcome a PR that fixed it as well.

**Comment by aftabnaveed** [comment]
@jakeboone02  you think it is a better idea to replace nanoid with short-uuid? I see the problem itself is not related to the query builder but nanonid. 

*UPDATE*
Ok I see you created a pull request, I will give the `short-uuid` branch a try and see how it goes.

**Comment by jakeboone02** [comment]
Great - let me know how it goes.

**Comment by jakeboone02** [comment]
I swapped out the id generator because I wanted to remove the hack requirement for IE11 support.  I think it has the added benefit of fixing SSR.

**Comment by aftabnaveed** [comment]
> I swapped out the id generator because I wanted to remove the hack requirement for IE11 support. I think it has the added benefit of fixing SSR.

I don't think it has, I still see this error 
```
crypto.getRandomValues() not supported. See https://github.com/uuidjs/uuid#getrandomvalues-not-supported Error: crypto.getRandomValues() not supported. See https://github.com/uuidjs/uuid#getrandomvalues-not-supported at a (/home/naveed/Projects/react-querybuilder/dist/index.js:6:26831) at E (/home/naveed/Projects/react-querybuilder/dist/index.js:6:32781) at Object.h (/home/naveed/Projects/react-querybuilder/dist/index.js:1:10901) at $ (/home/naveed/Projects/react-querybuilder/dist/index.js:6:10957) at q (/home/naveed/Projects/react-querybuilder/dist/index.js:6:10912) at QueryBuilder 
```
I don't know yet where the `crypto.getRandomValues` is being used. Looks like some other package is still causing it to crash.

**Comment by jakeboone02** [comment]
Looks like it's the `uuid` package, which `short-uuid` depends on.  Any ideas?  At this point I'm tempted to make my own id generator using `Math.random()`.

**Comment by aftabnaveed** [comment]
I am looking at a few issues reported here https://github.com/uuidjs/uuid/issues?q=getrandomvalues, let's see If I can get a clue.

**Comment by aftabnaveed** [comment]
Ok, I tried to install pollyfill for `crypto` but that did not work. I then went ahead and replaced `short-uuid` with `Math.random` as you suggested, but it still seems to be not working, however, the error has now changed to the following:

```
Invalid hook call. Hooks can only be called inside of the body of a function component. 
This could happen for one of the following reasons: 1. 
You might have mismatching versions of React and the renderer (such as React DOM) 2. 
You might be breaking the Rules of Hooks 3. 
You might have more than one copy of React in the same app See https://reactjs.org/link/invalid-hook-call for tips about how to debug and fix this problem. 
```

The `QueryBuilder.ts` is using hooks within the component but as per ReactJS documentation 
https://reactjs.org/warnings/invalid-hook-call-warning.html
They need to on top of the function

**Comment by jakeboone02** [comment]
The `QueryBuilder.tsx` file does not violate the Rules of Hooks.  It uses `useState` once and `useEffect` twice, all three times are at the top level of the `QueryBuilder` function.

**Comment by aftabnaveed** [comment]
Maybe I am not compiling or linking it properly, the node version I am using is `14.x`  and running `npm run dist`   to build the project and then using `npm link` to use it. Is there anything I could be missing here?

**Comment by jakeboone02** [comment]
I can't think of anything.  `react` is a peer dependency of `react-querybuilder`, so I don't think it would be pulling in its own version (which may cause the issue per option 3 above).  Are you sure you're not violating the Rules of Hooks in your own code?

**Comment by aftabnaveed** [comment]
> I can't think of anything. `react` is a peer dependency of `react-querybuilder`, so I don't think it would be pulling in its own version (which may cause the issue per option 3 above). Are you sure you're not violating the Rules of Hooks in your own code?

My code runs fine with the default repository, it simply doesn't work when I build it myself. I have been using ES6 for a while now but new to TypeScript and maybe that's the reason I can't figure it out

**Comment by aftabnaveed** [comment]
My code is quite straight forward and here it is: 

```javascript
const Deals = () => {
    
    let [fields, setFields] = useState({});
    let [fetched, setFetched] = useState(false);
    useEffect(() => {
        Api.get('/category-deals')
            .then((response) => {
                console.log(response);
                let newFields = response.data?.data?.content?.conditions.map((condition, index) => {
                    return {
                        label: condition.label,
                        name: condition.id,
                        valueEditorType: condition.id === 81 ? 'radio' : '',
                        values: condition.id == 81 ? [{ name: 'A', label: 'ABC' }, { name: 'C', label: 'CSD' }] : condition.id,
                        defaultValue: 'A'
                    }
                })

                setFields(newFields);
                setFetched(true);
            })
        
    }, []);

    const logQuery = (query) => {
        console.log(query)
    }
   return(
      <div className="with-bootstrap">
                                <Form className="form-inline">
                                    { ExecutionEnvironment.canUseDOM && fetched ?
                                        <QueryBuilder 
                                            controlClassnames={ bootstrap }
                                            controlElements={ { notToggle: NotToggle, valueEditor: ValueEditor } }
                                            fields={ fields } 
                                            onQueryChange={ logQuery } 
                                        
                                        />
                                    : <Spinner color="primary" />
                                    }
                                </Form>
                            </div>
  ) 
}
```

This works with the default npm package when I pull it from there but when I link it with my locally compiled package I get the above error.

**Comment by jakeboone02** [comment]
You're right, it looks pretty straight forward.  What if I go ahead and publish the `short-uuid` version of the package and you can use that?  That way we'd avoid any issues with `npm link`.

**Comment by aftabnaveed** [comment]
You mean to publish that to a local npm repository? Pardon my ignorance, this territory is fairly new for me

**Comment by jakeboone02** [comment]
No, I meant if I publish a new version to npm.  Version 3.9.1.

**Comment by aftabnaveed** [comment]
That probably won't fix the issue the problem is with the `uuid`  package which uses the `crypto` object not available in nodejs.

**Comment by jakeboone02** [comment]
Ah, you're right.  Maybe I'll just cancel that PR and create another one using `Math.random()` then.

**Comment by aftabnaveed** [comment]
Let's try that.

**Comment by jakeboone02** [comment]
OK, @aftabnaveed, I published v3.9.1 with the `Math.random()` change.  Let me know if that fixes your issue.

**Comment by aftabnaveed** [comment]
> OK, @aftabnaveed, I published v3.9.1 with the `Math.random()` change. Let me know if that fixes your issue.

That fixes the issue. Thank you very much!
